text/trace/115-add-sample-to-span-end.md

```diff
@@ -0,0 +1,204 @@
+fields such as start time in some context object, to be used to create a span
+later.
+
+However, most operations that involves I/O will need to defer span creation, as
+it is not possible to determine the appropriate span name until some data has
+been received.
+
+Examples include:
+
+* http(s) requests: the path and route on the server is only known after enough
+  of the payload is received to read the url.
+* RPC requests: RPCs typically include the desired operation in the
+  payload, after some bytes have been read.
+
+The span of the operation should have a start time that is right before the
+first byte is received.
+
+### Deferred span creation is worse than creating a span
+
+Ensuring a span is properly closed, even in error scenarios, is a desirable
+trait, and SDKs have ensured best-effort of closing of spans using language
+specific features such as exception handling or deferred close routines.
+
+Deferred span creation has not had as much development focus, which results in
+a less error handling that ensures that deferred spans are being created. This
+results in exceptional situations that occur before the span creation (e.g.
+invalid parsing of an http request) to not be represented in a span, as the
+deferred span is never created.
+
+A way to mitigate this would be to ensure the same level of error handling
+for deferred span creation as with spans themselves (create the span during a
+handled exception). But that would double the amount of maintenance to ensure
+robust span creation. Without span sampling as a concern, it would be desirable
+to just use a span instead, and focus on improving the robustness span creation.
+
+There is another side effect of deferred span creation, where there may be
+child spans that are created before the deferred span is. Examples here include
+a span to represent the receiving of data: the span that represents receiving data
+would have started before the name of the span is available, since that depends
+on receiving and parsing the incoming payload. The result will at worst be
+multiple root spans that should actually be children of the deferred span, since
+there will be no active span present when the children are created.
+
+### It is better to address the issues with sampling
+
+Given the need for an almost identical, most likely less robust code path
+for deferred spans, and the fact that it is a very common pattern, solving
+issues with sampling would ensure a more robust and maintainable code base.
+
+## Explanation
+
+## Additional Hooks
+
+A sampler would consume the same arguments it does today:
+
+- span name
+- span context
+- span kind
+- span attributes
+- span links
+
+The sampler would incorporate additional hooks, one to match each modification
+of events on an existing span:
+
+* onUpdateName, called by updateName
+* onSetAttribute, called by setAttribute
+* onAddEvent, called by addEvent
+* onSetStatus, called by setStatus
+* onEnd, called at the end of the span
```

Comment:
Is calling `onEnd` equivalent to `Alternative: Always sample at the end`?

Reply:
yes, although the alternative is lesser featured (since there's fewer hooks to update sampling decisions).